sensor/admission-control/manager/manager_impl.go

```diff
@@ -150,22 +149,22 @@ func (m *manager) Sync(ctx context.Context) error {
 	case m.syncC <- &syncSig:
 	case <-ctx.Done():
 		return ctx.Err()
-	case <-m.stoppedSig.Done():
-		return m.stoppedSig.ErrorWithDefault(pkgErr.New("manager was stopped"))
+	case <-m.stopper.Client().Stopped().Done():
+		return m.stopper.Client().Stopped().ErrorWithDefault(pkgErr.New("manager was stopped"))
 	}
 
 	select {
 	case <-syncSig.Done():
 		return nil
 	case <-ctx.Done():
 		return ctx.Err()
-	case <-m.stoppedSig.Done():
-		return m.stoppedSig.ErrorWithDefault(pkgErr.New("manager was stopped"))
+	case <-m.stopper.Client().Stopped().Done():
+		return m.stopper.Client().Stopped().ErrorWithDefault(pkgErr.New("manager was stopped"))
 	}
 }
 
 func (m *manager) Start() error {
-	if !m.stopSig.Reset() {
+	if m.stopper.LowLevel().ResetStopRequest() {
```

Comment:
There was a bug in this line before.
It was
```go
if !m.stopSig.Reset() {
```
but it should have been
```go
if m.stopSig.Reset() {
```

The reason it still worked was because `m.stopSig` wasn't initialized! Check that `NewManager` does not set it.

That's the reason it's better to use interfaces instead of structs. You can use uninitialized struct and have weird effects but you can't use `nil` interface.

Reply:
Good catch on the issue! 

However, I'm struggling to understand why is this needed in the first place. It seems like this manager's `Start()` is only called once on admission control's `main` function: https://github.com/stackrox/stackrox/blob/master/sensor/admission-control/main.go#L79-L80

If the component is just starting, in which case would the signal be already triggered during startup? 

Additionally, the function call and the error message are even more confusing. If the reset was successful (i.e. `ResetStopRequest()` returns `true`) it returns and error saying the manager has already started. AFAIU this would mean that the manager was _stopped_ instead. The only reason I can see this would return true is if the component started, received a stop signal, and the component was requested to start again.

Reply:
You're right, the `Start()` is called right after creating a `New()` manager. It is impossible that something calls `Stop()` on `mgr` between those lines.

Either the code is over-designed or there _was_ some different usage in the past.

Regarding `ResetStopRequest()` return value, what you write here

>  The only reason I can see this would return true is if the component started, received a stop signal, and the component was requested to start again.

matches my understanding.

There is no way of identifying whether the manager got **started** from `stopper` (currently) or `stopSig` previously.

The only thing I can offer in this case is to delete the entire `if-error` block to make the code less confusing. However, I prefer to not take this too far as there are **many** simple and difficult bugs in the code I touched and I don't want to be fixing everything for everyone.

Reply:
I'm fine leaving the code as is. Just wanted to point out. 

Maybe add a comment linking to this discussion? Just in case someone finds this code in the future and feels brave enough to try removing the if and testing if anything breaks?

Reply:
Sorry, I wasn't clear enough. I meant that I will fix this issue (the fix is simple and I just committed it) but don't want to get on a path of fixing all issues because some might be more involved.